src/h3/_cy/memory.pyx

```diff
@@ -0,0 +1,115 @@
+from libc cimport stdlib
+from cython.view cimport array
+from .h3lib cimport H3int
+
+
+cdef size_t move_nonzeros(H3int* a, size_t n):
+    """ Move nonzero elements to front of array `a` of length `n`.
+    Return the number of nonzero elements.
+
+    Loop invariant: Everything *before* `i` or *after* `j` is "done".
+    Move `i` and `j` inwards until they equal, and exit.
+    You can move `i` forward until there's a zero in front of it.
+    You can move `j` backward until there's a nonzero to the left of it.
+    Anything to the right of `j` is "junk" that can be reallocated.
+
+    | a | b | 0 | c | d | ... |
+            ^           ^
+            i           j
+
+
+    | a | b | d | c | d | ... |
+            ^       ^
+            i       j
+    """
+    cdef:
+        size_t i = 0
+        size_t j = n
+
+    while i < j:
+        if a[j-1] == 0:
+            j -= 1
+            continue
+
+        if a[i] != 0:
+            i += 1
+            continue
+
+        # if we're here, we know:
+        # a[i] == 0
+        # a[j-1] != 0
+        # i < j
+        # so we can swap! (actually, move a[j-1] -> a[i])
+        a[i] = a[j-1]
+        j -= 1
+
+    return i
+
+
+cdef H3int[:] empty_memory_view():
+    # todo: get rid of this?
+    # there's gotta be a better way to do this...
+    # create an empty cython.view.array?
+    cdef:
+        H3int a[1]
+
+    return (<H3int[:]>a)[:0]
+
+
+cdef class H3MemoryManager:
```

Comment:
I'd expect a memory manager to have broader responsibilities for managing allocations (have more verbs essentially); this seems to be more like a wrapped buffer or a block of managed memory?

Reply:
Yeah, basically just an object to be responsible for deallocating the memory before that responsibility is handed off to a `memoryview` object.

Open to different names :)

Reply:
Can we make this a context manager? Seems like a good fit unless for some reason a context manager is hard/slow in Cython

Reply:
I was also thinking about using a context manager here. I forget the specific examples, but it seemed like not every situation naturally fell into a context manager pattern.

An object that guarantees memory clean up would, at least, be an improvement over what we have now. I wouldn't be opposed to refactoring that into a context manager in the future.

Either way, I'll keep it in mind as I play with this. And if a context manager does end up working out, maybe I'll do that.